[202405][Rebase&&FF] Firmware Performance

Flickdm opened the pull request:
This PR adds a series of commits intended to add firmware performance measurements to a platform.

### Overview of commits:

1. MdePkg: Add EFI_PRE_READY_TO_BOOT_GUID and EFI_POST_READY_TO_BOOT_GUID
   1. This commit adds two new GUIDS required for Delay Dispatch
2. REBASE: Revert "MdePkg: Added header file for Delayed Dispatch PPI"
   1. This commit reverts a commit brought in from upstream that modifies the DelayDispatch.h file in MdePkg that was moved to MdeModulePkg
 3. MdeModulePkg/MdeModulePkg.dec: Adds Delay Dispatch
    1. This commit adds Delay Dispatch
 4. MdeModulePkg: Adds Performance Measurement V2
    1. This commit adds the initial performance measurements
 5. Perf-instrument PreReadyToBoot, ReadyToBoot, PostReadyToBoot signals …
    1. This commit adds the signals required for performance measurements
 6. Merged PR 5181: Change FirmwarePerformanceDxe.c EndofDxeEvent tpl to …
    1. This commit changes the TPL_NOTIFYs to TPL_CALLBACKS
 7. MdeModulePkg: PeiMain: Adds performance measurement for PeiMain
    1. Adds peformance measurements for PEI


- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

----
### How This Was Tested
From 2311.

dmitryantipov committed:
MdeModulePkg: Adds Performance Measurement V2

This combines the following commits:
https://github.com/microsoft/mu_basecore/commit/25a1aba511
https://github.com/microsoft/mu_basecore/commit/831dde508c
https://github.com/microsoft/mu_basecore/commit/7a38833

dmitryantipov committed:
Perf-instrument PreReadyToBoot, ReadyToBoot, PostReadyToBoot signals and callbacks

Perf-instrument PreReadyToBoot, ReadyToBoot, and PostReadyToBoot signals and callbacks

antklein committed:
Merged PR 5181: Change FirmwarePerformanceDxe.c EndofDxeEvent tpl to TPL_CALLBACK

Change FirmwarePerformanceDxe.c EndofDxeEvent tpl to TPL_CALLBACK and
Change DxeCorePerformanceLib.c EndofDxeEvent tpl to TPL_CALLBACK

makubacki commented on MdeModulePkg/Core/Dxe/Dispatcher/Dispatcher.c:
```diff
@@ -414,6 +414,7 @@ CoreDispatcher (
     //
     // If the dispatcher is running don't let it be restarted.
     //
+    PERF_FUNCTION_END (); // MS_CHANGE
```
Please try to replace these with `MU_CHANGE` if you can before publishing the PR.

os-d commented on MdePkg/Include/Guid/EventGroup.h:
```diff
@@ -49,4 +49,15 @@ extern EFI_GUID  gEfiEventDxeDispatchGuid;
 extern EFI_GUID  gEfiEndOfDxeEventGroupGuid;
 
+// MU_CHANGE [BEGIN]
+#define EFI_PRE_READY_TO_BOOT_GUID \
```
Can you split this into two commits, first for MdePkg, second for MdeModulePkg?

os-d commented on MdePkg/Include/Ppi/DelayedDispatch.h:
```diff
@@ -1,102 +0,0 @@
-/** @file
```
I assume you need this commit to build for changes? Seems unrelated otherwise.

Flickdm replied:
Accurate - having two copies causes the build to fail.
I could split this into it's own smaller PR. The problem is that Edk2 has this in MdePkg and we have it in MdeModulePkg

os-d commented on MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c:
```diff
@@ -5,6 +5,8 @@ Copyright (c) 2006 - 2024, Intel Corporation. All rights reserved.<BR>
 (C) Copyright 2016 Hewlett Packard Enterprise Development LP<BR>
 SPDX-License-Identifier: BSD-2-Clause-Patent
 
+Copyright (c) 2017, Microsoft Corporation.
```
Copyright (C) Microsoft Corporation.

Please ensure any additions have this format.

os-d commented on MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c:
```diff
@@ -1411,6 +1413,24 @@ PeiDispatcher (
+        Status = PeiServicesNotifyPpi (&mDelayedDispatchNotifyDesc);
+        ASSERT_EFI_ERROR (Status);
+      }
+    }                                                                    // MS_CHANGE End
```
While you are making changes, can you change the MS_CHANGEs to MU_CHANGEs? Also, the end tag should go on the next brace down.

os-d commented on MdeModulePkg/Core/Pei/PeiMain.h:
```diff
@@ -316,7 +343,6 @@ struct _PEI_CORE_INSTANCE {
 ///
 #define PEI_CORE_INSTANCE_FROM_PS_THIS(a) \
   CR(a, PEI_CORE_INSTANCE, Ps, PEI_CORE_HANDLE_SIGNATURE)
-
```
This can be removed.

os-d commented on MdePkg/MdePkg.dec:
```diff
@@ -903,6 +903,25 @@
   # GUID used to provide initrd to linux via LoadFile2 protocol
```
Can you also split this commit per pkg?

os-d commented on MdeModulePkg/Core/Pei/Dispatcher/Dispatcher.c:
```diff
@@ -1660,13 +1988,16 @@ PeiDispatcher (
               PeimEntryPoint (PeimFileHandle, (const EFI_PEI_SERVICES **)PeiServices);
-              // PERF_END (PeiServices, L"PEIM", PeimFileHandle, 0);
+              // PERF_END (PeiServices, L"PEIM", PeimFileHandle, 0); // MU_CHANGE
+              // PERF_ENTRY_POINT_END (PeimFileHandle); // MU_CHANGE
```
We can drop these commented additions?

os-d commented on MdeModulePkg/Library/UefiBootManagerLib/InternalBm.h:
```diff
@@ -68,7 +68,6 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Library/DxeServicesLib.h>
 #include <Library/ReportStatusCodeLib.h>
 #include <Library/CapsuleLib.h>
-#include <Library/PerformanceLib.h>
```
Can we do the canonical style of commenting out with a MU_CHANGE tag?

os-d commented on MdeModulePkg/Library/UefiBootManagerLib/UefiBootManagerLib.inf:
```diff
@@ -59,7 +59,6 @@
   MemoryAllocationLib
   DxeServicesLib
   ReportStatusCodeLib
-  PerformanceLib
```
Can we do the canonical style of commenting out with a MU_CHANGE tag?

Flickdm committed:
MdePkg: Add EFI_PRE_READY_TO_BOOT_GUID and EFI_POST_READY_TO_BOOT_GUID

This patch adds two new GUIDs to the EventGroup.h file. These GUIDs are
    - EFI_PRE_READY_TO_BOOT_GUID
    - EFI_POST_READY_TO_BOOT_GUID

This change was split off the commit:
https://github.com/microsoft/mu_basecore/commit/7a38833
